core/src/main/java/org/apache/iceberg/SnapshotParser.java

```diff
@@ -129,13 +129,12 @@ static Snapshot fromJson(JsonNode node) {
           "Cannot parse summary from non-object value: %s",
           sNode);
 
+      operation = JsonUtil.getString(OPERATION, sNode);
       ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
       Iterator<String> fields = sNode.fieldNames();
       while (fields.hasNext()) {
         String field = fields.next();
-        if (field.equals(OPERATION)) {
-          operation = JsonUtil.getString(OPERATION, sNode);
-        } else {
+        if (!field.equals(OPERATION)) {
```

Comment:
I disagree with this change. Although the operation is required by the spec, we do not want to fail to read a table if it isn't there. We can be more permissive than the spec.

@kevinjqliu can you open a PR to revert this please?

Reply:
@rdblue Sorry, not sure I completely follow. If it's been required in the spec to have an operation defined in the case a summary mapping is defined, an Iceberg compliant implementation would've always populated this correctly when writing out the metadata.

Is there a historical reason where we expect some tables not to have this populated? I'd wonder how that'd work because validation during commits needs to check against the operation for intermediate snapshots, so we would be somehow handling "unknown" operation cases (which I guess the default would be to assume conflict because it's unknown)

Reply:
I went through https://github.com/apache/iceberg-python/issues/1106, seems like another engine is now producing the metadata without the operation key but that just doesn't seem spec compliant. 
I guess there's the argument that we want to prevent needless failures but I actually think this may be needed (the validation example I mentioned above, unless I'm missing something?)

Reply:
Created a revert PR, #11409.

> Although the operation is required by the spec, we do not want to fail to read a table if it isn't there. We can be more permissive than the spec.

I agree that we want to be more permissive than the spec when reading a table. However, what should we do about working with a table that is not compliant with the spec? We would need to add extra logic to handle V2 tables without the `operation` field. 

> Is there a historical reason where we expect some tables not to have this populated? 

From looking at the [commit history for `SnapshotParser.toJson`](https://github.com/apache/iceberg/blame/377cab742adacc18f1c81d372bb4d65a549e180e/core/src/main/java/org/apache/iceberg/SnapshotParser.java#L63-L66), the `summary` and `operation` fields either both exist or are both missing. So historically, the Java Iceberg implementation should not produce one without the other.



Here's my current understanding:
- V1 tables do not have `operation` and `summary` fields. 
- V2 tables should have both `operation` and `summary` fields
- Another engine generated a V2 table with the `summary` field but without the `operation` field. This is technically not compliant with the V2 spec. We want to be more permissive and still allow reading this non-compliant table. But we should figure out how to work with this V2 table that is without Snapshot operation.